### PR TITLE
TSCBasic: protect against an invalid API usage

### PR DESCRIPTION
**Pull request by compnerd**
`filePathRepresentation` may not be invoked on an empty string.  This
protects against that case which does occur during the SPM test suite
execution.

### CODE REVIEW

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test macOS platform

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test macOS platform

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Review by neonichu — APPROVED** [review]

**Comment by tomerd** [comment]
this does a bit more than just the empty string case, can you elaborate on what is going on here?

cc @abertelrud

File: Sources/TSCBasic/Path.swift [hunk 1]
```diff
@@ -254,7 +259,18 @@ public struct RelativePath: Hashable {
     /// normalization or canonicalization.  This will construct a path without
     /// an anchor and thus may be invalid.
     fileprivate init(unsafeUncheckedPath string: String) {
-        self.init(PathImpl(string: string))
+        if string.isEmpty {
```
Comment (tomerd) [inline Sources/TSCBasic/Path.swift:262]: does the case need to be a guard and then return a ~const empty representation?

Comment (tomerd) [inline Sources/TSCBasic/Path.swift:262]: does the case need to be a guard and then return a ~const empty representation?

**Comment by compnerd** [comment]
@swift-ci please test macOS platform

**Comment by tomerd** [comment]
@compnerd this does a bit more than just the empty string case, can you elaborate on what is going on here / why all the thing manipulation is needed?

cc @abertelrud @milseman

**Comment by tomerd** [comment]
@compnerd this does a bit more than just the empty string case, can you elaborate on what is going on here / why all the thing manipulation is needed?

cc @abertelrud @milseman

**Comment by compnerd** [comment]
@tomerd I don't follow, its doing nothing at all ...

`NSString.fileSystemRepresentation` cannot be invoked on an empty string, so the check (or guard if you really want it just spelt that way for your tastes) is just protecting against that.

If it is not empty, we normalize the string and return that because the path operations should be performed on the normalized form (otherwise we need to go in at each point where an API is called and re-normalize the path before each call, which is extremely expensive as strings must be re-encoded to UTF16, have the conversion, re-encoded to UTF8, manipulated and then re-encoded to UTF16 to perform the operation).

**Comment by compnerd** [comment]
Hmm, wait, do you mean the `self.init(PathImpl(string: string))` in the case that the string is empty?  I don't see how that is any different than writing `self.init(PathImpl(string: ""))`.  We already know that `string` is empty, so its just passing in `string` as opposed to creating a new String value from the `""` literal.  Is that you prefer spelling that as `self.init(Self.constant)`?

**Comment by tomerd** [comment]
iiuc the non-empty code path is now:

```
#if _runtime(_ObjC)
             self.init(PathImpl(string: String(cString: NSURL(fileURLWithPath: string).fileSystemRepresentation)))
 #else
             let normalized: UnsafePointer<Int8> = string.fileSystemRepresentation
             defer { normalized.deallocate() }

             self.init(PathImpl(string: String(cString: normalized)))
 #endif
```

that is quite different from that the code used to be, and the motivation for this change (why do we need to use fileSystemRepresentation? what problem is it solving) is not explained as the motivation for this PR

**Comment by compnerd** [comment]
Ah, I see.

The difference in the ObjC runtime is because `fileSystemRepresentation` behaviour is different.  The reason for that is to give similar behaviour across the two paths.

As to why the call to `.fileSystemRepresentation` is to canonicalise the path (that will rewrite the filepath to use the natural path separator rather than POSIX.  This is important because subsequent use of the resulting path makes assumptions that the path is in the canonicalised form rather than POSIX.  If we do not perform this at this point, future attempts to get pieces of the path fail.

**Comment by tomerd** [comment]
> As to why the call to .fileSystemRepresentation is to canonicalise the path (that will rewrite the filepath to use the natural path separator rather than POSIX. This is important because subsequent use of the resulting path makes assumptions that the path is in the canonicalised form rather than POSIX. If we do not perform this at this point, future attempts to get pieces of the path fail.

can you give some concrete examples of what this does exactly / what concrete problem it solves. I am a bit worried about replying on fileSystemRepresentation. 

@milseman  @abertelrud can you chime in here too?

**Comment by milseman** [comment]
> @milseman @abertelrud can you chime in here too?

Sorry, I'm not familiar with what this is. This doesn't look to be using `FilePath`, should it be?

**Comment by compnerd** [comment]
Right, this is still using the path utilities from tools-support-core, not the swift-system FilePath.

**Comment by tomerd** [comment]
@compnerd iirc https://github.com/apple/swift-tools-support-core/pull/369 and https://github.com/apple/swift-package-manager/pull/5910 are designed to avoid some of these issues, we should prioritize getting them over the finish lines so we need less workarounds like this

**Comment by compnerd** [comment]
@tomerd are you okay with merging this for now?  I'm going to likely take a pause on some of the SPM work and see if we can get the two mentioned changes merged as that is going to just be generally easier to make progress with.

**Comment by tomerd** [comment]
@compnerd are you asking if we can merge https://github.com/apple/swift-tools-support-core/pull/369 and https://github.com/apple/swift-package-manager/pull/5910? I am happy to merge them if you can confirm they are not breaking anything for windows and helping with some of the windows path issues

**Comment by compnerd** [comment]
@tomerd I meant merging this change so that I can have a little bit of breathing room to work through those two changes :)

**Comment by compnerd** [comment]
@tomerd okay, can you please at least get the patch pair that you have building on Windows?  I should be able to put it through its paces on Windows once you have the patches complete.  Can we get that done early today so that I can test it?  Alternatively, can we merge this temporarily so that we can work through the issues in the much larger refactoring?

**Comment by compnerd** [comment]
I believe that this can be closed off now that we have the other change merged.

**Pull request closed without merging**